src/pixelbasher.cpp

```diff
@@ -0,0 +1,166 @@
+//
+//
+// Copyright the mso-test contributors
+//
+// SPDX-License-Identifier: MPL-2.0
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include <algorithm>
+#include <cmath>
+
+#include "pixel.hpp"
+#include "pixelbasher.hpp"
+
+BMP PixelBasher::compare_bmps(const BMP &original, const BMP &target, bool enable_minor_differences)
+{
+	int min_width = std::min(original.get_width(), target.get_width());
+	int min_height = std::min(original.get_height(), target.get_height());
+	int pixel_stride = 4;
+
+	BMP diff(original);
+
+	std::vector<bool> intersection_mask = get_intersection_mask(original, target, min_width, min_height);
+	// The diff data is based on the base image, so we create a new data vector to hold the modified pixels
+	auto &original_data = original.get_data();
+	auto &target_data = target.get_data();
+	std::vector<uint8_t> diff_data = original.get_data();
+
+	int original_width = original.get_width();
+	int target_width = target.get_width();
+	// Loops through the min width and height, if size of images differ slightly.
+	for (int y = 0; y < min_height; y++)
+	{
+		for (int x = 0; x < min_width; x++)
+		{
+			// The intersection mask is based on height and width not stride, so we calculate it separately
+			int original_index = (y * original_width + x) * pixel_stride;
+			int target_index = (y * target_width + x) * pixel_stride;
+			int mask_index = y * min_width + x;
+
+			Pixel original_pixel = Pixel::get_pixel(original_data, original_index);
+			Pixel target_pixel = Pixel::get_pixel(target_data, target_index);
+			bool near_edge = intersection_mask[mask_index];
+
+			std::vector<uint8_t> bgra = compare_pixels(original_pixel, target_pixel, diff, near_edge, enable_minor_differences);
```

Comment:
The following code uniformly operates on the individual members of `bgra`, so it is arguably better to use a `vector` than to use `Pixel` with its individual `blue`, `green`, `red`, `alpha` members.  But `vector` is rather "fat" and "expensive" (and doesn't say that it always contains four elements), so a `std::array<uint8_t, pixel_stride>` would arguably be an even better type here.